oauth/client_factory.go

```diff
@@ -40,9 +48,14 @@ type ClientFactoryConfig struct {
 	ApiServerCAPath string
 }
 
-// CreateUserAuthClient creates a new client based on the provided configuration. We use this client for k8s requests with user's token (e.g.: SelfSubjectAccessReview).
+type ClientFactory struct {
+	FactoryConfig ClientFactoryConfig
+	HTTPClient    rest.HTTPClient
+}
+
+// CreateUserAuthClientForNamespace creates a new client based on the provided configuration. We use this client for k8s requests with user's token (e.g.: SelfSubjectAccessReview).
 // Note that configuration is potentially modified during the call.
-func CreateUserAuthClient(args *ClientFactoryConfig) (AuthenticatingClient, error) {
+func (f ClientFactory) CreateUserAuthClientForNamespace(ctx context.Context, namespace string) (AuthenticatingClient, error) {
```

Comment:
What about caching the clients? If I remember correctly the Kubernetes client contains an internal object cache which we are probably re-creating each time we create a new client.
We should probably cache per workspace, not per namespace though, because that would mean that we'd have several clients for the same cluster, each having its own object cache.